### PR TITLE
[RFC] Inclusion of Caffeine Applet

### PR DESCRIPTION
**Pull request by yursan9**
This is port of my python [Caffeine Applet](https://github.com/yursan9/budgie-caffeine-applet).

### CODE REVIEW

File: src/applets/caffeine/CaffeineWindow.vala [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+        var icon = event_box.get_child ();
+        event_box.remove (icon);
+
+        var time = timer.get_value_as_int();
+        if (mode.active)
+        {
+            // Fetch power settings default value
+            fetch_default ();
+
+            timer.sensitive = false;
```
Comment (JoshStrobl) [inline src/applets/caffeine/CaffeineWindow.vala]: ` timer.sensitive` setting should just be an inverse value of `mode.active` and be outside the if/else block IMO

File: src/applets/caffeine/CaffeineWindow.vala [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+            // Add timeout callback if timer's spinbox is not 0
+            if (time > 0)
+            {
+                Timeout.add_seconds(time * 60, on_timer_out, Priority.DEFAULT);
+            }
+        }
+        else
+        {
+            timer.sensitive = true;
```
Comment (JoshStrobl) [inline src/applets/caffeine/CaffeineWindow.vala]: See other remark on `timer.sensitive`

File: src/applets/caffeine/CaffeineWindow.vala [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+            // Fetch power settings default value
+            fetch_default ();
+
+            timer.sensitive = false;
+            session_settings.set_uint ("idle-delay", 0);
+            power_settings.set_boolean ("idle-dim", false);
+            power_settings.set_string ("sleep-inactive-ac-type", "nothing");
+            power_settings.set_string ("sleep-inactive-battery-type", "nothing");
+
+            icon = new Gtk.Image.from_icon_name("caffeine-cup-full", Gtk.IconSize.MENU);
```
Comment (JoshStrobl) [inline src/applets/caffeine/CaffeineWindow.vala]: icon setting can use a ternary operator on the string, along the lines of (but not necessarily representative of working code :D):

```
icon = new Gtk.Image.from_icon_name("caffeine-cup-" + (mode.active ? "full" : "empty"), Gtk.IconSize.MENU);
```

This and the event_box add call could be done outside of the if/else block.

File: src/applets/caffeine/CaffeineWindow.vala [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+        }
+        else
+        {
+            timer.sensitive = true;
+            session_settings.set_uint ("idle-delay", default_idle_delay);
+            power_settings.set_boolean ("idle-dim", default_idle_dim);
+            power_settings.set_string ("sleep-inactive-ac-type", default_sleep_inactive_ac_type);
+            power_settings.set_string ("sleep-inactive-battery-type", default_sleep_inactive_battery_type);
+
+            icon = new Gtk.Image.from_icon_name("caffeine-cup-empty", Gtk.IconSize.MENU);
```
Comment (JoshStrobl) [inline src/applets/caffeine/CaffeineWindow.vala]: See remarks on other icon setting.

**Comment by yursan9** [comment]
Thank you Josh! I'll work on the changes today

**Comment by JoshStrobl** [comment]
LGTM, can you get this commit squashed into a single one? I'll merge it after :smile:

**Comment by yursan9** [comment]
Sorry, Josh! Don't merge the branch yet. I put wrong commit in there, somehow...

**Pull request merged**